reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,147 @@
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val schedulerJob = SupervisorJob()
+
+    private val scope = CoroutineScope(schedulerJob + dispatcher + CoroutineExceptionHandler { _, throwable ->
+        RxJavaPlugins.onError(throwable)
+    })
+
+    override fun scheduleDirect(block: Runnable): Disposable =
+        scheduleDirect(block, 0, TimeUnit.MILLISECONDS)
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (!scope.isActive) return Disposables.disposed()
+        val newBlock = RxJavaPlugins.onSchedule(block)
+        return scope.launch {
+            if (delay > 0) {
+                delay(unit.toMillis(delay))
+            }
+            newBlock.run()
+        }.asDisposable()
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, schedulerJob)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerJob = SupervisorJob(parentJob)
+        private val workerScope = CoroutineScope(workerJob + dispatcher)
+        private val blockChannel = Channel<SchedulerChannelTask>(Channel.UNLIMITED)
+
+        init {
+            workerScope.launch {
+                while (isActive) {
+                    val task = blockChannel.receive()
+                    task.execute()
+                }
+            }
+        }
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable =
+            schedule(block, 0, TimeUnit.MILLISECONDS)
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            if (!workerScope.isActive) return Disposables.disposed()
+
+            val newBlock = RxJavaPlugins.onSchedule(block)
+
+            fun addTaskToQueue(job: Job) {
+                val task = SchedulerChannelTask(newBlock, job)
+                blockChannel.offer(task)
+            }
+
+            val job = if(delay == 0L) {
+                Job(workerJob).also {
+                    addTaskToQueue(it)
+                }
+            } else {
+                workerScope.launch {
+                    delay(unit.toMillis(delay))
+                    val task = SchedulerChannelTask(newBlock, checkNotNull(coroutineContext[Job]))
+                    blockChannel.offer(task)
```

Comment:
These two lines should be a call to `addTaskToQueue`. I think you also can't use this `Job`, since the job will become completed as soon as this `launch` lambda returns. Instead, just pull the `Job(workerJob)` call in the first branch out into a val and pass that same instance to `addTaskToQueue` in both branches.

So just to be clear:
```kotlin
val taskJob = Job(workerJob)

fun addTaskToQueue() {
  val task = SchedulerChannelTask(newBlock, taskJob)
  blockChannel.offer(task)
}

if (delay == 0L) {
  addTaskToQueue()
} else {
  // Use `taskJob` as the parent here so the delay will also get cancelled if the Disposable
  // is disposed.
  workerScope.launch(taskJob) {
    // Delay *before* enqueuing the task, so other tasks (e.g. via schedule without delay)
    // aren't blocked by the delay.
    delay(unit.toMillis(delay))
    // Once the task is ready to run, it still needs to be executed via the queue to comply
    // with the Scheduler contract of running all worker tasks in a non-overlapping manner.
    addTaskToQueue()
  }
}

return taskJob.asDisposable()
```
Including those comments, since the implications of this code are evidently pretty subtle, and it would be helpful to future readers/maintainers to explain why it's done the way it is.